File: exercise/php/day21/src/Tour/TourCalculator.php
<?php

namespace Tour;

class TourCalculator
{
    private $steps;
    private $deliveryTime = 0;
    private $calculated = false;

    public function __construct($steps)
    {
        $this->steps = $steps;
    }

    public function isCalculated()
    {
        return $this->calculated;
    }

    public function getDeliveryTime()
    {
        return $this->deliveryTime;
    }

    public function calculate()
    {
        if (empty($this->steps)) {
            return "No locations !!!";
        } else {
            $result = '';

            usort($this->steps, function ($a, $b) {
                return $a->time <=> $b->time;
            });

            foreach ($this->steps as $step) {
                if (!$this->calculated) {
                    $this->deliveryTime += $step->deliveryTime;
                    $result .= $this->formatLine($step, $this->deliveryTime) . PHP_EOL;
                }
            }

            $hhMmSs = gmdate("H:i:s", $this->deliveryTime);
            $result .= "Delivery time | {$hhMmSs}" . PHP_EOL;
            $this->calculated = true;

            return $result;
        }
    }

    private function formatLine($step, $deliveryTime)
    {
        if ($step === null) {
            throw new \InvalidArgumentException();
        } else {
            return "{$step->time} : {$step->label} | {$step->deliveryTime} sec";
        }
    }
}

$steps = [
    new Step('08:00', 'Start', 0),
    new Step('08:30', 'Location A', 1800),
    new Step('09:00', 'Location B', 1800),
];

$calculator = new TourCalculator($steps);
$result = $calculator->calculate();

if (strpos($result, 'No locations') !== false) {
    echo "Error: {$result}\n";
} else {
    echo $result;
}

?>

File: exercise/java/day21/src/main/java/tour/TourCalculator.java
package tour;

import io.vavr.collection.List;
import io.vavr.control.Either;

import java.time.Duration;
import java.util.Comparator;

import static io.vavr.control.Either.left;
import static io.vavr.control.Either.right;

public class TourCalculator {
    private List<Step> steps;
    private boolean calculated = false;
    private double deliveryTime = 0;

    public TourCalculator(List<Step> steps) {
        this.steps = steps;
    }

    public boolean isCalculated() {
        return calculated;
    }

    public double getDeliveryTime() {
        return deliveryTime;
    }

    public Either<String, String> calculate() {
        if (steps == null || steps.isEmpty()) {
            return left("No locations !!!");
        } else {
            StringBuilder result = new StringBuilder();

            for (Step s : steps.sorted(Comparator.comparing(Step::time))) {
                if (!calculated) {
                    this.deliveryTime += s.deliveryTime();
                    result.append(fLine(s, deliveryTime)).append("\n");
                }
            }

            String str = formatDurationToHHMMSS(Duration.ofSeconds((long) this.deliveryTime));
            result.append("Delivery time | ").append(str);
            calculated = true;

            return right(result.toString());
        }
    }

    private static String formatDurationToHHMMSS(Duration duration) {
        long hours = duration.toHours();
        long minutes = duration.toMinutesPart();
        long seconds = duration.toSecondsPart();

        return String.format("%02d:%02d:%02d", hours, minutes, seconds);
    }

    private String fLine(Step step, double x) {
        if (step == null)
            throw new IllegalStateException();
        else
            return step.time() + " : " + step.label() + " | " + step.deliveryTime() + " sec";
    }
}